### PR TITLE
Rewrite in ES6

### PR DESCRIPTION
**Pull request by paulcbetts**
This PR cleans up this project into a proper ES6 package

## TODO:

- [x] Port the source code over
- [x] Port the tests over
- [x] Verify on Windows
- [x] Verify on OS X
- [ ] Fix up the README

### CODE REVIEW

**Comment by kevinsawicki** [comment]
Thanks for this @paulcbetts, #1 was just merged so conflicts abound :hammer:

**Comment by paulcbetts** [comment]
@lukeapage So, I kinda rewrote this without reading #1 first, but what do you think about this:

* Port back the use of Jetpack like #1, this library looks great
* Use `debug` instead of the custom logging framework, I think it's more standard and lets people turn on/off noisy output

**Comment by lukeapage** [comment]
Sounds good to me.  I'd advise against the await keyword since its not yet stage 4 and wont be in es2016..  still better to have an unfinished standard than coffeescript though imo.

**Comment by kevinsawicki** [comment]
@paulcbetts please ping me when this is ready for review :eyes:

**Comment by paulcbetts** [comment]
@kevinsawicki Will look at this tomorrow

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    return exe;
+  }
+
+  let target = path.join('.', exe);
+  if (statSyncNoException(target)) {
+    d(`Found executable in currect directory: ${target}`);
+    return target;
+  }
+
+  let haystack = process.env.PATH.split(isWindows ? ';' : ':');
```
Comment (rbkreisberg) [inline src/index.js]: node has a nice helper for PATH delimiter.

https://nodejs.org/api/path.html#path_path_delimiter

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    return null;
+  }
+}
+
+function locateExecutableInPath(exe) {
+  // NB: Windows won't search PATH looking for executables in spawn like
+  // Posix does
+
+  // Files with any directory path don't get this applied
+  if (exe.match(/[\\\/]/)) {
```
Comment (rbkreisberg) [inline src/index.js]: node has a helper for path separators.

https://nodejs.org/api/path.html#path_path_sep

**Comment by paulcbetts** [comment]
@kevinsawicki I'm gonna finish this today then you can have an :eyeglasses:

File: spec/.eslintrc [hunk 3]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (lukeapage) [inline spec/.eslintrc]: `.eslintrc` is deprecated, use `.eslintrc.json` - http://eslint.org/docs/user-guide/configuring#configuration-file-formats

Comment (lukeapage) [inline spec/.eslintrc]: Noticed this is eslint 1 - maybe worth getting it in and then doing this and updating to eslint v2 in a seperate PR

File: src/spawn-promise.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+      let chunk = b.toString();
+      stdout += chunk;
+    };
+
+    proc.stdout.on('data', bufHandler);
+    proc.stderr.on('data', bufHandler);
+    proc.on('error', (e) => reject(e));
+
+    proc.on('close', (code) => {
+      // NB: Close fires before stdout fully flushes (i.e. you'll get 'data'
```
Comment (kevinsawicki) [inline src/spawn-promise.js]: Can you wait until the streams themselves emit `close` for this then?

That seems possibly safer than a 2 second timeout which could be flaky.

Comment (kevinsawicki) [inline src/spawn-promise.js]: Meaning `proc.stdout.on('close')` and `proc.stderr.on('close')`

Comment (paulcbetts) [inline src/spawn-promise.js]: Ooh, yes I can

File: src/index.js (hunk 2 above)
Comment (paulcbetts) [inline src/index.js]: Let's just let this roll for now

File: spec/.eslintrc (hunk 3 above)
Comment (paulcbetts) [inline spec/.eslintrc]: Yeah, let's do this in another PR

**Comment by paulcbetts** [comment]
@kevinsawicki This should be :gem:

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+import jetpack from 'fs-jetpack';
+import fs from 'fs';
+
+const d = require('debug')('electron-windows-installer:main');
+
+temp.track();
+
+export function p(strings, ...values) {
+  let newPath = String.raw(strings, ...values);
+  let parts = _.map(newPath.split(/[\\\/]/), (x) => x || '/');
```
Comment (havenchyk) [inline src/index.js]: Why not to use `newPath.split(/[\\\/]/).map` here? @paulcbetts

Comment (paulcbetts) [inline src/index.js]: Because I don't know the Array prototype methods!

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  let defaults = {
+    description: appMetadata.description || '',
+    exe: `${appMetadata.name}.exe`,
+    iconUrl: 'https://raw.githubusercontent.com/atom/electron/master/atom/browser/resources/win/atom.ico',
+    title: appMetadata.productName || appMetadata.name
+  };
+
+  let metadata = _.assign({}, appMetadata, options, defaults);
```
Comment (havenchyk) [inline src/index.js]: Maybe use Object.assign instead?

File: src/index.js (hunk 5 above)
Comment (havenchyk) [inline src/index.js]: Is it a joke? I can create a separate PR to reduce using of lodash.

**Comment by havenchyk** [comment]
Guys, how're things going here? Do you have plans to review it, @kevinsawicki?
Is it a final version, @paulcbetts?

**Comment by kevinsawicki** [comment]
Published in 2.0.0, thanks for this @paulcbetts
